src/transactions/SetTrustLineFlagsOpFrame.cpp

```diff
@@ -0,0 +1,195 @@
+                                                   TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mSetTrustLineFlags(mOperation.body.setTrustLineFlagsOp())
+{
+}
+
+ThresholdLevel
+SetTrustLineFlagsOpFrame::getThresholdLevel() const
+{
+    return ThresholdLevel::LOW;
+}
+
+bool
+SetTrustLineFlagsOpFrame::isVersionSupported(uint32_t protocolVersion) const
+{
+    return protocolVersion >= 16;
+}
+
+bool
+SetTrustLineFlagsOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    ZoneNamedN(applyZone, "SetTrustLineFlagsOp apply", true);
+
+    {
+        LedgerTxn ltxSource(ltx); // ltxSource will be rolled back
+        auto header = ltxSource.loadHeader();
+        auto sourceAccountEntry = loadSourceAccount(ltxSource, header);
+        auto const& sourceAccount = sourceAccountEntry.current().data.account();
+
+        const uint32_t authFlags =
+            AUTHORIZED_FLAG | AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG;
+
+        bool authNotRevocable = !(sourceAccount.flags & AUTH_REVOCABLE_FLAG);
+
+        // This condition covers all three revoke scenarios (AUTHORIZED_FLAG ->
+        // AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG (You need to clear
+        // AUTHORIZED_FLAG here, if not, then the auth flags will be invalid and
+        // the op will return SET_TRUST_LINE_FLAGS_INVALID_STATE),
+        // AUTHORIZED_FLAG -> 0, or AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG ->
+        // 0).
+        // If setFlags and clearFlags overlap, the op would've failed during
+        // validation
+        bool clearingAnyAuth =
+            mSetTrustLineFlags.clearFlags &&
+            (*mSetTrustLineFlags.clearFlags & authFlags) != 0;
+        bool settingAuthorized =
+            mSetTrustLineFlags.setFlags &&
+            (*mSetTrustLineFlags.setFlags & AUTHORIZED_FLAG) != 0;
+        if (authNotRevocable && clearingAnyAuth && !settingAuthorized)
+        {
+            innerResult().code(SET_TRUST_LINE_FLAGS_CANT_REVOKE);
+            return false;
+        }
+    }
+
+    auto key =
+        trustlineKey(mSetTrustLineFlags.trustor, mSetTrustLineFlags.asset);
+    bool shouldRemoveOffers = false;
+    uint32_t expectedFlagValue = 0;
+    {
+        // trustline is loaded in this inner scope because it can be loaded
+        // again in removeOffersByAccountAndAsset
+        auto trust = ltx.load(key);
+        if (!trust)
+        {
+            innerResult().code(SET_TRUST_LINE_FLAGS_NO_TRUST_LINE);
+            return false;
+        }
+
+        expectedFlagValue = trust.current().data.trustLine().flags;
+        if (mSetTrustLineFlags.clearFlags)
+        {
+            expectedFlagValue &= ~*mSetTrustLineFlags.clearFlags;
+        }
+
+        if (mSetTrustLineFlags.setFlags)
+        {
+            expectedFlagValue |= *mSetTrustLineFlags.setFlags;
+        }
+
+        shouldRemoveOffers =
+            isAuthorizedToMaintainLiabilities(trust) && expectedFlagValue == 0;
```

Comment:
This doesn't seem right to me. What if `expectedFlagValue == TRUSTLINE_CLAWBACK_ENABLED_FLAG`?

Reply:
Wow good catch. I'll add a test for this. Good reminder that the CAP-0035 changes still need to be tested thoroughly.

Reply:
Fixed and added a testcase.